File: c++/146. LRU Cache.cpp
class LRUCache {
public:
    LRUCache(int capacity):size(capacity) {
    }
    
    int get(int key) {
        auto it = hash.find(key);
        if (it != hash.end()) {
            cache.splice(cache.begin(), cache, it->second); //splice 从cache中去除迭代器指向的元素，添加到cache头部
            return it->second->second;
        }
        return -1;
    }
    
    void put(int key, int value) {
        auto it = hash.find(key);
        if (it != hash.end()) {
            it->second->second = value;
            cache.splice(cache.begin(), cache, it->second);
            return;
        }
        
        cache.insert(cache.begin(), make_pair(key, value));
        hash[key] = cache.begin();
        if (cache.size() > size) {
            hash.erase(cache.back().first);
            cache.pop_back();
        }
        
    }

private:
    unordered_map<int, list<pair<int, int>>::iterator> hash;
    list<pair<int, int>> cache;
    int size;
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: python/146. LRU Cache.py
#!usr/bin/env python  
#-*- coding:utf-8 _*-  

class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.cache = {}
        self.que = []
        

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.cache:
            return -1
        # 更新优先级
        idx = self.que.index(key)
        self.que.pop(idx)
        self.que.insert(0, key)
        return self.cache[key]
        

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        self.cache[key] = value
        if key in self.que:
            idx = self.que.index(key)
            self.que.pop(idx)
        self.que.insert(0, key)    
        if len(self.que) > self.capacity:
            lastKey = self.que[-1]
            del self.cache[lastKey]
            self.que.pop()


        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)